### PR TITLE
CAMEL-10791-added option to handle empty directories while unzipping

### PR DESCRIPTION
**Pull request by onders86**

### CODE REVIEW

File: camel-core/src/main/java/org/apache/camel/component/file/GenericFileProducer.java [hunk 1]
```diff
@@ -274,7 +274,11 @@ public void writeFile(Exchange exchange, String fileName) throws GenericFileOper
             log.trace("About to write [{}] to [{}] from exchange [{}]", new Object[]{fileName, getEndpoint(), exchange});
         }
 
-        boolean success = operations.storeFile(fileName, exchange);
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/component/file/GenericFileProducer.java]: We do NOT want changes in the core file producer. Try to make changes only to the camel-zipfile component

Comment (onders86) [inline camel-core/src/main/java/org/apache/camel/component/file/GenericFileProducer.java]: i believe you are right. i did not want such change IMHO as well. there reason why i went there was camel-zipfile was passing either ZipIterator or Outputstream to producer. For the case in the issue, iterator usage is required for Splitter EIP and Message is created with FILENAME header and conveyed to producer and core file producer does all the thing like creating directories and and files on the producer endpoint. What i did in camel-zipfile component to create a Message with FILENAME header <path> + "." which only indicated the target empty directory to be created. And core file producer already does manupulation like creating paths and files on producer endpoint. I just added a simple condition to bypass storeFile method call and directory creation for empty folders is OK. that was the purpose. I have not come up with an easier solution  and i did not feel like i would do something dirty on core file producer, therefore, i chose that path. If this does not make any sense, i can close the PR.

Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/component/file/GenericFileProducer.java]: Yeah then please close

File: components/camel-zipfile/src/main/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormat.java [hunk 2]
```diff
@@ -72,6 +73,11 @@ public void marshal(final Exchange exchange, final Object graph, final OutputStr
 
     @Override
     public Object unmarshal(final Exchange exchange, final InputStream inputStream) throws Exception {
+        if (supportEmptyDirectory) {
```
Comment (zhurlik) [inline components/camel-zipfile/src/main/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormat.java]: Maybe a single line will be more beautiful :)
exchange.getIn().setHeader("unzipEmptyDirectorySupported", isSupportEmptyDirectory());

**Comment by onders86** [comment]
ok removed core file producer update and only edited component. thanks

File: components/camel-zipfile/src/test/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormatTest.java [hunk 3]
```diff
@@ -187,6 +217,35 @@ public void setUp() throws Exception {
         deleteDirectory(TEST_DIR);
         super.setUp();
     }
+
```
Comment (zhurlik) [inline components/camel-zipfile/src/test/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormatTest.java]: try-with-resources will simplify this code:

`
    private static void copy(File file, OutputStream out) throws IOException {
        try (InputStream in = new FileInputStream(file)) {
            copy(in, out);
        }
    }
    private static void copy(InputStream in, File file) throws IOException {
        try (OutputStream out = new FileOutputStream(file)) {
            copy(in, out);
        }
    }
`

Comment (onders86) [inline components/camel-zipfile/src/test/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormatTest.java]: i am a bit old fashioned. fine suggestion. i did you way

File: components/camel-zipfile/src/main/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormat.java [hunk 4]
```diff
@@ -72,6 +73,7 @@ public void marshal(final Exchange exchange, final Object graph, final OutputStr
 
     @Override
     public Object unmarshal(final Exchange exchange, final InputStream inputStream) throws Exception {
+        exchange.getIn().setHeader("unzipEmptyDirectorySupported", isSupportEmptyDirectory());
         if (usingIterator) {
             return new ZipIterator(exchange.getIn());
```
Comment (davsclaus) [inline components/camel-zipfile/src/main/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormat.java]: Do not use a Camel message header for this, but set the option on the iterator using getter/setter

    ZipIterator it = new ...
    it.setUnzipEmptyDirectory(...)

File: components/camel-zipfile/src/main/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormat.java [hunk 5]
```diff
@@ -39,6 +39,7 @@
  */
 public class ZipFileDataFormat extends ServiceSupport implements DataFormat, DataFormatName {
     private boolean usingIterator;
+    private boolean supportIteratorForEmptyDirectory;
```
Comment (davsclaus) [inline components/camel-zipfile/src/main/java/org/apache/camel/dataformat/zipfile/ZipFileDataFormat.java]: Can we rename this to `allowEmptyDirectory' to make it shorter and use similar name we have done for other such option

**Comment by davsclaus** [comment]
You also need to add the new option in the model in camel-core, eg

https://github.com/apache/camel/blob/master/camel-core/src/main/java/org/apache/camel/model/dataformat/ZipFileDataFormat.java#L38

**Comment by davsclaus** [comment]
Thanks that looks fine.

I wonder if you would do a similar functionality to camel-tarfile? I think its similar to this module, and could maybe also support empty folders.

**Comment by davsclaus** [comment]
Thanks for the PR it has been merged. Do you mind closing this?

**Comment by onders86** [comment]
yes sure i'll look into camel-tarfile as well. thanks. closing

**Pull request closed without merging**